File: kwebui/frontend/static/js/widgets/container.js
"use strict";

registerWidget("container", {
  props: ["data"],
  computed: {
    // Mirrors image.js's widthStyle -- see there for why this is
    // `width: max-content` rather than `align-self: flex-start` (it's
    // what lets a *parent* container's `horizontal_alignment` reposition
    // a nested, unstretched container instead of that nested container's
    // own alignSelf overriding it).
    widthStyle() {
      if (this.data.props.stretch) return { width: "100%" };
      const width = this.data.props.width;
      return width > 0 ? { width: `${width}px` } : { width: "max-content" };
    },
    // No stretch-equivalent or ambient-default to escape here, unlike
    // widthStyle: a flex column's main axis (height) is never stretched
    // by an ambient parent default the way the cross axis (width) is, so
    // -1/0 (the default) simply omits the style key and the container
    // sizes to its content, no `height: max-content` workaround needed.
    heightStyle() {
      const height = this.data.props.height;
      return height > 0 ? { height: `${height}px` } : {};
    },
    // flex-direction: row for direction="horizontal" -- children laid out
    // side by side at their own natural size, rather than stacked (the
    // "vertical" default, which omits the key so the class rule's own
    // `flex-direction: column` applies -- byte-identical to before this
    // prop existed). wrap only really matters combined with horizontal,
    // but is passed straight through as flex-wrap regardless: CSS accepts
    // it either way and there's no reason for this widget to police the
    // combination itself.
    directionStyle() {
      const style = {};
      if (this.data.props.direction === "horizontal") style.flexDirection = "row";
      if (this.data.props.wrap) style.flexWrap = "wrap";
      return style;
    },
    // This container's own align-items/justify-content, positioning ITS
    // children -- unrelated to widthStyle above, which is about how THIS
    // container itself is sized/positioned within its own parent. Maps
    // the user-facing left/right/top/bottom words to their flexbox
    // equivalents; left/right isn't valid CSS for align-items (only
    // flex-start/flex-end are). Omitting the key entirely (rather than
    // mapping a null/unset alignment to "stretch"/"flex-start" explicitly)
    // is what keeps the default byte-identical to before this prop
    // existed -- see create()'s docstring in container.py for why that
    // matters (alert/listbox/textedit/slider all lean on the inherited
    // default).
    //
    // horizontal_alignment/vertical_alignment keep their *meaning*
    // ("along the horizontal/vertical axis") regardless of direction, but
    // which flex property that maps to (align-items vs. justify-content)
    // depends on which axis is currently the flex main axis -- swapped
    // here for direction="horizontal", where the main axis is horizontal
    // instead of the column default's vertical.
    alignStyle() {
      const style = {};
      const h = { left: "flex-start", center: "center", right: "flex-end" }[this.data.props.horizontal_alignment];
      const v = { top: "flex-start", center: "center", bottom: "flex-end" }[this.data.props.vertical_alignment];
      const horizontalDirection = this.data.props.direction === "horizontal";
      if (h) style[horizontalDirection ? "justifyContent" : "alignItems"] = h;
      if (v) style[horizontalDirection ? "alignItems" : "justifyContent"] = v;
      return style;
    },
    // null/undefined (the default -- Python's None) means "use the
    // theme's own default", so it omits the key entirely and lets
    // .sg-container-body's own CSS rule (padding: var(--sg-container-
    // padding-vertical) var(--sg-container-padding-horizontal), see
    // base.css) apply -- an inline style here always wins over that rule
    // regardless of specificity, which is exactly what makes an explicit
    // vertical_padding/horizontal_padding "overwrite the theme". 0 is a
    // real, distinct value from null: it's an explicit "no padding"
    // override, not "defer to the theme". Each axis is independent, so
    // e.g. only vertical_padding set still lets horizontal fall through
    // to its own theme variable in the same inline shorthand.
    paddingStyle() {
      const v = this.data.props.vertical_padding;
      const h = this.data.props.horizontal_padding;
      if (v == null && h == null) return {};
      const vv = v == null ? "var(--sg-container-padding-vertical)" : `${v}px`;
      const hh = h == null ? "var(--sg-container-padding-horizontal)" : `${h}px`;
      return { padding: `${vv} ${hh}` };
    },
  },
  // heightStyle/alignStyle apply to the inner .sg-container-body div, not
  // this fieldset -- see base.css's comment on .sg-container for why: a
  // <fieldset display:flex> whose own parent is also flex (always true
  // here) silently breaks justify-content, so the fieldset stays a plain
  // non-flex box (border/legend/width only) and the actual flex-column
  // layout for children lives one level down, on a div whose parent
  // (this fieldset) is never itself flex.
  template: `
    <fieldset
      class="sg-container"
      :style="widthStyle"
      :data-border="!!data.props.border"
      :data-rounded="!!data.props.border_roundness"
    >
      <legend v-if="data.props.caption" class="sg-container-caption">{{ data.props.caption }}</legend>
      <div class="sg-container-body" :style="[heightStyle, directionStyle, alignStyle, paddingStyle]">
        <slot></slot>
      </div>
    </fieldset>
  `,
});

File: kwebui/widgets/container.py
"""Container widget: a generic region whose entire content can be
cleared and rebuilt -- used to swap between different "pages" of widgets
while the app keeps running (see ``examples/workflow.py``).

    page = app.container()
    page.text("Page 1")
    ...
    page.clear()          # wipe it
    page.text("Page 2")   # rebuild with different content

Structurally the same append-only container idea as
``ColumnWidget``/``SidebarWidget``/``TopbarWidget`` (same ``__getattr__``
dynamic factory, duplicated rather than shared -- matches the existing
precedent across those modules), plus a ``clear()`` method mirroring
``EmptyWidget.clear()``. The difference from ``Empty`` is exactly that:
``Empty`` holds a single child and replacing it means calling another
``slot.<widget>(...)``, whereas a ``Container`` is meant to hold a whole
group of widgets at once, so swapping its content is "clear(), then
rebuild with several calls" rather than a single replacing call.
"""

from __future__ import annotations

from typing import Any, Callable

from ..events import Event
from ..plugin import WidgetPlugin
from ..widget import Widget


class ContainerWidget(Widget):
    def set_width(self, width: float) -> "ContainerWidget":
        """Resize the container. -1 or 0 falls back to sizing to its
        content. Ignored while ``stretch`` is True."""
        self.update(width=width)
        return self

    def set_height(self, height: float) -> "ContainerWidget":
        """Give the container a fixed height. -1 or 0 falls back to sizing
        to its content, like ``width``'s default -- there is no ``stretch``
        equivalent for height. A fixed height is what makes
        ``vertical_alignment`` visible (see ``create()``'s docstring):
        without one, the container never has more height than its content
        needs, so there's nothing to center within."""
        self.update(height=height)
        return self

    def set_stretch(self, stretch: bool) -> "ContainerWidget":
        """Toggle whether the container fills its parent's width."""
        self.update(stretch=stretch)
        return self

    def set_border(self, border: bool) -> "ContainerWidget":
        """Toggle the container's border."""
        self.update(border=border)
        return self

    def set_border_roundness(self, border_roundness: bool) -> "ContainerWidget":
        """Toggle whether the container's border corners are rounded."""
        self.update(border_roundness=border_roundness)
        return self

    def set_caption(self, caption: str) -> "ContainerWidget":
        """Change the caption shown breaking the top border line.
        Pass "" to remove it."""
        self.update(caption=caption)
        return self

    def set_horizontal_alignment(self, horizontal_alignment: str | None) -> "ContainerWidget":
        """Change how children are aligned along the container's width.
        One of "left"/"center"/"right", or None to go back to filling the
        full width (the default -- see ``create()``'s docstring)."""
        self.update(horizontal_alignment=horizontal_alignment)
        return self

    def set_vertical_alignment(self, vertical_alignment: str | None) -> "ContainerWidget":
        """Change how children are packed along the container's height.
        One of "top"/"center"/"bottom", or None for the default (top --
        only visible once the container has more height than its content
        needs, e.g. given an explicit height by its own parent)."""
        self.update(vertical_alignment=vertical_alignment)
        return self

    def set_shortkey(self, shortkey: str | None) -> "ContainerWidget":
        """Change (or clear, with ``None``) the keyboard shortcut that
        calls ``on_keypress`` -- see ``create()``'s docstring."""
        self.update(shortkey=shortkey)
        return self

    def set_on_keypress(self, on_keypress: Callable[[], None] | None) -> "ContainerWidget":
        """Change (or clear, with ``None``) the callback ``shortkey`` calls."""
        self.update(on_keypress=on_keypress)
        return self

    def set_vertical_padding(self, vertical_padding: float | None) -> "ContainerWidget":
        """Change the space (in px) between the container's top/bottom
        edge and its content, or go back to the theme's own default
        (``--sg-container-padding-vertical`` in ``base.css``) with
        ``None``."""
        self.update(vertical_padding=vertical_padding)
        return self

    def set_horizontal_padding(self, horizontal_padding: float | None) -> "ContainerWidget":
        """Change the space (in px) between the container's left/right
        edge and its content, or go back to the theme's own default
        (``--sg-container-padding-horizontal`` in ``base.css``) with
        ``None``."""
        self.update(horizontal_padding=horizontal_padding)
        return self

    def __getattr__(self, name: str) -> Any:
        if name.startswith("_"):
            raise AttributeError(name)
        app = self.__dict__.get("_app")
        if app is None or name not in app.registry.names():
            raise AttributeError(name)

        def factory(*args: Any, **kwargs: Any) -> Widget:
            child = app.registry.get(name).create(app._next_id(), *args, **kwargs)
            # Parent+broadcast before wiring up `_app` -- see EmptyWidget's
            # factory for why the order matters (self-broadcasting widgets
            # like alert/toast/popup/columns would otherwise announce
            # themselves unparented first and get mounted twice).
            self.children.append(child)
            if self._app is not None:
                self._app._on_widget_changed(self)
            child._app = app
            return child

        return factory

    def clear(self) -> "ContainerWidget":
        """Remove all current children -- call before rebuilding with
        different content (e.g. switching to a different "page")."""
        self.children = []
        if self._app is not None:
            self._app._on_widget_changed(self)
        return self


class ContainerPlugin(WidgetPlugin):
    """
    Example:
        page = app.container()
        page.text("Hello")

        card = app.container(width=300, border=True, border_roundness=True)
        wide = app.container(stretch=True, border=False)
        titled = app.container(caption="Settings")  # caption breaks the top border line
        centered = app.container(width=400, horizontal_alignment="center")
        panel = app.container(height=300, vertical_alignment="center")  # a fixed-height "panel"
        panel.button("Centered")  # ... vertically centered inside it
        app.container(shortkey="ctrl+k", on_keypress=open_search)  # fires while this container is visible
        padded = app.container(vertical_padding=16, horizontal_padding=24)  # breathing room around content
    """

    widget_name = "container"

    def create(
        self,
        widget_id: str,
        *,
        width: float = -1,
        height: float = -1,
        stretch: bool = False,
        border: bool = True,
        border_roundness: bool = True,
        caption: str = "",
        horizontal_alignment: str | None = None,
        vertical_alignment: str | None = None,
        shortkey: str | None = None,
        on_keypress: Callable[[], None] | None = None,
        vertical_padding: float | None = None,
        horizontal_padding: float | None = None,
    ) -> ContainerWidget:
        """``horizontal_alignment`` ("left"/"center"/"right") and
        ``vertical_alignment`` ("top"/"center"/"bottom") position children
        within the container's own box, once that box is bigger than its
        content -- e.g. a fixed ``width`` wider than the children, or a
        fixed ``height`` taller than the children.

        Left at the default ``None``, children keep today's original
        behavior: stretched to the container's full width and packed at
        the top, exactly as if this parameter didn't exist. This is a
        deliberate choice, not an oversight -- several widgets with a
        visible border/background (``success()``/``info()``/``warning()``/
        ``error()``, ``listbox``, ``textedit``, ``slider``) rely on that
        implicit full-width stretch from their parent to reach their own
        max-width, not on any width of their own. Picking an explicit
        ``horizontal_alignment`` shrinks every child to its own natural
        size first, *then* aligns it -- so a `listbox`/`textedit`/`slider`
        inside a horizontally-aligned container renders narrower than in
        an unaligned one.

        ``height`` (like ``width``) defaults to ``-1``, sizing to content;
        there is no ``stretch``-for-height equivalent. Unlike
        ``horizontal_alignment``, ``vertical_alignment`` has no
        default-preserving caveat to worry about: a container's height is
        intrinsic (sized to its content) regardless of what
        `vertical_alignment` is set to, so it stays invisible until you
        also give the container a fixed ``height`` (or it's otherwise
        handed extra height by its own parent) -- at which point it
        actually has more room than its content needs, and something to
        center/bottom-align within.

        ``shortkey`` (e.g. ``"k"``, ``"shift+k"``, ``"ctrl+k"``,
        ``"shift+ctrl+k"`` -- modifiers in any order, joined with ``+``)
        binds a keyboard combo that calls ``on_keypress``, following the
        exact same rules as `button`'s `shortkey` (see its docstring):
        global listener, but only fires while this container is actually
        visible on the page (not `.hide()`-den, not inside a hidden
        ancestor), and bare/shift-only combos are suppressed while typing
        in a text field. Unlike `button`, a container has no `on_click` to
        reuse, so it gets its own dedicated `on_keypress` callback instead.

        ``vertical_padding``/``horizontal_padding`` (in px) add space
        between the container's own edge and its content -- top/bottom
        and left/right respectively. Both default to ``None``, which
        means "use the theme's own default" (``--sg-container-padding-
        vertical``/``--sg-container-padding-horizontal`` in ``base.css``,
        `10px` each out of the box) rather than a hardcoded number here --
        restyle the theme variables to change every container's default
        padding at once, or pass an explicit value to override just this
        one instance (pass `0` for none at all, distinct from `None`
        which defers to the theme). Padding is applied inside the border,
        so it doesn't move the border/caption themselves, only the room
        between the border and what's inside.
        """
        props = {
            "width": width,
            "height": height,
            "stretch": stretch,
            "border": border,
            "border_roundness": border_roundness,
            "caption": caption,
            "horizontal_alignment": horizontal_alignment,
            "vertical_alignment": vertical_alignment,
            "shortkey": shortkey,
            "on_keypress": on_keypress,
            "vertical_padding": vertical_padding,
            "horizontal_padding": horizontal_padding,
        }
        return ContainerWidget(widget_id, self.widget_name, props)

    def handle_event(self, widget: Widget, event: Event) -> None:
        # Mirrors ButtonPlugin's "shortkey" handling, but with no "click"
        # equivalent to reuse -- a container's shortkey has always meant
        # on_keypress, there's no separate "real" interaction it aliases.
        if event.type == "shortkey":
            callback = widget.props.get("on_keypress")
            if callback is not None:
                callback()
